.github/workflows/cabal.yml

```diff
@@ -13,41 +13,39 @@ jobs:
         cabal: ["3.6"]
         ghc:
           - "8.10.7"
-          - "9.2.2"
+          - "9.2.3"
 
     steps:
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@v3
         with:
           submodules: true
 
-      - uses: haskell/actions/setup@v1
-        name: Setup GHC and cabal-install
+      - name: Setup z3
+        uses: pavpanchekha/setup-z3@1.2.2
         with:
-          ghc-version: ${{ matrix.ghc }}
-          cabal-version: ${{ matrix.cabal }}
-
-      - uses: rwe/actions-hlint-setup@v1
-        name: Set up HLint
-        with:
-          version: "3.3.4"
+          version: "4.8.7"
 
-      - uses: actions/cache@v2
-        name: cache ~/.cabal/store
+      - name: Setup GHC and cabal-install
+        uses: haskell/actions/setup@v2
         with:
-          path: ~/.cabal/store
-          key: ${{ runner.os }}-${{ matrix.ghc }}-cabal
+          ghc-version: ${{ matrix.ghc }}
+          cabal-version: ${{ matrix.cabal }}
 
-      - uses: pavpanchekha/setup-z3@1.2.2
-        name: setup z3
+      - name: Cache ~/.cabal/packages, ~/.cabal/store and dist-newstyle
+        uses: actions/cache@v3
         with:
-          version: "4.8.7"
+          path: |
+            ~/.cabal/packages
+            ~/.cabal/store
+            dist-newstyle
+          key: ${{ runner.os }}-${{ matrix.ghc }}-cabal-${{ hashFiles('**/*.cabal', '**/cabal.project', '**/cabal.project.freeze') }}
+          restore-keys: ${{ runner.os }}-${{ matrix.ghc }}-cabal-
```

Comment:
I was just noticing that perhaps this is going to cause caches to grow indefinitely as the dependencies change, right?

Would it be preferable to just drop `restore-keys` and start the cache afresh if there is not an exact key match?

Reply:
We could drop `restore-keys` as it is only used when there's a cache miss, see [using the cache action](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#example-using-the-cache-action).

Reply:
> I was just noticing that perhaps this is going to cause caches to grow indefinitely as the dependencies change, right?

> GitHub will remove any cache entries that have not been accessed in over 7 days. There is no limit on the number of caches you can store, but the total size of all caches in a repository is limited to 10 GB.
> FROM: [usage limits and eviction policy](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy).